File: flutter-backend/src/models/rebalanceModel.js
import mongoose from 'mongoose';

const searchResultSchema = new mongoose.Schema({
  symbol: { type: String, index: true },
  name: String,
  type: { type: String, default: 'EQUITY' },
  sector: String,
  exchange: String,
  updatedAt: { type: Date, default: Date.now }
});

const rebalanceRecSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', index: true },
  symbol: String,
  action: { type: String, enum: ['BUY', 'SELL', 'HOLD'], default: 'HOLD' },
  quantity: Number,
  reason: String,
  createdAt: { type: Date, default: Date.now }
});

export default mongoose.model('RebalanceRecommendation', rebalanceRecSchema);

File: Frontend/lib/models/rebalance_model.dart
enum RebalanceAction { buy, sell, hold }

class RebalanceRecommendation {
  final String symbol;
  final String name;
  final double currentWeight;
  final double targetWeight;
  final double amount;
  final RebalanceAction action;
  final String reason;

  RebalanceRecommendation({
    required this.symbol,
    required this.name,
    required this.currentWeight,
    required this.targetWeight,
    required this.amount,
    required this.action,
    required this.reason,
  });

  String get actionString {
    switch (action) {
      case RebalanceAction.buy:
        return 'BUY';
      case RebalanceAction.sell:
        return 'SELL';
      case RebalanceAction.hold:
        return 'HOLD';
    }
  }

  String get formattedAmount => '\$${amount.toStringAsFixed(2)}';
  String get formattedPercentage =>
      '${(targetWeight - currentWeight).toStringAsFixed(1)}%';

  double get difference => (targetWeight - currentWeight).abs();
}